### PR TITLE
Get eval function wrapper

### PR DESCRIPTION
**Pull request by ravishk007**
* Implemented the get eval wrapper which takes input as model_name and dataset_name and map it to the task_type. It is using namedtuple.
* Refactored segmentation eval function in the src directory to give metrics dictionary directly.
* Used the subclassing of registry class
* Refactored wrapper.py
* Implemented normalisation function for dataset name
* Added unit test for get eval function test_evaluation_functions.py
* Refactored test_pretrained_model_accuracy.py for get_eval_function

### CODE REVIEW

**Comment by ivan-lazarevich** [comment]
@ravishk007 please fix the pylint issues:

Trailing whitespace (trailing-whitespace)
Too many return statements (7/6) (too-many-return-statements)
Unnecessary "elif" after "return" (no-else-return)

File: deeplite_torch_zoo/src/objectdetection/eval/coco/coco_evaluator.py [hunk 1]
```diff
@@ -156,3 +156,26 @@ def yolo_eval_coco(model, data_root, gt=None, device="cuda",
     with torch.no_grad():
         return YoloCOCOEvaluator(model, dataset, gt=gt, net=net,
             img_size=img_size, progressbar=progressbar).evaluate()
+
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='yolo', dataset_name='car_detection')
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/src/objectdetection/eval/coco/coco_evaluator.py:160]: minor comment: please add a space before `model_name` for code style

File: deeplite_torch_zoo/src/objectdetection/eval/lisa_eval.py [hunk 2]
```diff
@@ -136,7 +136,7 @@ def evaluate(self):
 
         return ap  # Average Precision  (AP) @[ IoU=050 ]
 
-
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='yolo', dataset_name='lisa')
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/src/objectdetection/eval/lisa_eval.py:139]: same as above

Comment (ivan-lazarevich) [reply deeplite_torch_zoo/src/objectdetection/eval/lisa_eval.py:139]: also I suggest renaming `model_name` -> `model_type`, `dataset_name` -> `dataset_type` for the evaluator registry args

File: deeplite_torch_zoo/src/objectdetection/eval/voc/voc_evaluator.py [hunk 3]
```diff
@@ -153,6 +154,11 @@ def __calc_APs(self, iou_thresh=0.5, use_07_metric=False):
         return APs
 
 
+
+
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='yolo', dataset_name='person_pet_vehicle_detection')
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='yolo', dataset_name='person_detection')
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='yolo', dataset_name='voc')
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/src/objectdetection/eval/voc/voc_evaluator.py:161]: introducing `dataset_type` would eliminate the need to register multiple types. all the datasets here would have the same `voc` type

File: deeplite_torch_zoo/src/objectdetection/eval/voc/voc_evaluator.py [hunk 4]
```diff
@@ -175,3 +181,13 @@ def yolo_eval_voc(
         result["mAP"] = mAP
 
     return result
+
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='yolo', dataset_name='voc07')
+def yolo_voc07_eval(
+    model, data_root, num_classes=20, device="cuda", net="yolo3",
+    img_size=448, is_07_subset=True, progressbar=False, **kwargs
+):
+    result = yolo_eval_voc(model, data_root, num_classes=20, device="cuda", net="yolo3",
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/src/objectdetection/eval/voc/voc_evaluator.py:190]: minor: you can return `yolo_eval_voc(..)` w/o the temp var

File: deeplite_torch_zoo/src/segmentation/fcn/eval.py [hunk 5]
```diff
@@ -21,4 +23,5 @@ def evaluate_fcn(model, loader, device="cuda"):
                 label_trues.append(lt)
                 label_preds.append(lp)
     metrics = label_accuracy_score(label_trues, label_preds, n_class)
-    return metrics[1]
+    miou = {"miou":metrics[1] }
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/src/segmentation/fcn/eval.py:26]: minor: code style `miou = {"miou": metrics[1]}`

File: deeplite_torch_zoo/utils/registry.py [hunk 6]
```diff
@@ -75,3 +75,31 @@ def get(self, model_name, dataset_name):
...
+
+    def register(self, task_type, model_name, dataset_name):
+        def _register(obj_name, obj):
+            if obj_name in self._registry_dict:
+                raise KeyError(f'{obj_name} is already registered in the model wrapper registry')
+            self._registry_dict[obj_name] = obj
+
+        def wrap(obj):
+            cls_name = model_name
+            #self._registry_key.__new__.__defaults__ = (None,) * len(self._registry_key._fields)
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/utils/registry.py:93]: you can remove this line and set default values to args of the `register` method -> no need to supply None during registration to args that are None (e.g. dataset_type)

File: deeplite_torch_zoo/wrappers/eval/rcnn.py [hunk 7]
```diff
@@ -1,15 +1,17 @@
...
-
+from deeplite_torch_zoo.wrappers.registries import EVAL_WRAPPER_REGISTRY
 __all__ = ["rcnn_eval_coco"]
 
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='fasterrcnn_resnet50_fpn', dataset_name='coco')
 def rcnn_eval_coco(model, data_loader, gt=None, device="cuda", net="rcnn"):
     model.to(device)
     with torch.no_grad():
-        return RCNNCOCOEvaluator(
+        mAP = RCNNCOCOEvaluator(
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/wrappers/eval/rcnn.py:11]: is this change needed?

File: deeplite_torch_zoo/wrappers/eval/ssd.py [hunk 8]
```diff
@@ -11,16 +11,20 @@
...
 from deeplite_torch_zoo.src.objectdetection.ssd.repo.vision.ssd.mobilenet_v2_ssd_lite import create_mobilenetv2_ssd_lite_predictor
-
+from deeplite_torch_zoo.wrappers.registries import EVAL_WRAPPER_REGISTRY
 
 __all__ = [
 			"vgg16_ssd_eval_func",
 			"mb1_ssd_eval_func",
 			"mb2_ssd_eval_func",
 			"mb2_ssd_lite_eval_func",
+            "mb2_ssd_coco_eval_func",
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/wrappers/eval/ssd.py:21]: minor: fix code style (identation)

File: deeplite_torch_zoo/wrappers/eval/ssd.py [hunk 9]
```diff
@@ -11,16 +11,20 @@
...
 		]
-
-
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='vgg16_ssd', dataset_name='wider_face')
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='resnet50_ssd', dataset_name='voc')
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='resnet34_ssd', dataset_name='voc')
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='resnet18_ssd', dataset_name='voc')
+@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection',model_name='vgg16_ssd', dataset_name='voc')
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/wrappers/eval/ssd.py:27]: same as above: you should introduce `model_type="ssd"` and `dataset_type=None` (which would be by default)

this would change into smth like: 

```
@EVAL_WRAPPER_REGISTRY.register(task_type='object_detection', model_type='ssd')
def vgg16_ssd_eval_func(
    model, data_loader, iou_threshold=0.5, use_2007_metric=True, device="cuda"
):
```

File: deeplite_torch_zoo/wrappers/models/objectdetection/ssd.py [hunk 10]
```diff
@@ -29,7 +29,7 @@
     "mb2_ssd_coco_gm_6": "mb2-ssd-coco-6-0_455-1ce14b4728801c61.pth",
     "mb2_ssd_voc_20": "mb2-ssd-voc-20-0_440-8402e1cfc3e076b1.pth",
     "vgg16_ssd_voc_20": "vgg16-ssd-voc-mp-0_7726-b1264e8beec69cbc.pth",
-    "vgg16_ssd_wider_face_8": "vgg16-ssd-wider_face-0_707-8c76d36acb083648.pth",
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/wrappers/models/objectdetection/ssd.py:32]: plz keep the dataset names according to the "{name}_{num_classes}" pattern. u can use `dataset_type="wider_face"` for the registry while the name stays "wider_face_1" (i guess 8 was a mistake)

File: deeplite_torch_zoo/wrappers/wrapper.py [hunk 11]
```diff
@@ -65,6 +84,29 @@ def get_model_by_name(
     model = model_func(pretrained=pretrained, progress=progress, device=device)
     return model.half() if fp16 else model
 
+def get_eval_function(model_name="", dataset_name=""):
+    
+    _register_key = namedtuple('RegistryKey', ['model_name', 'dataset_name'])
+    key = _register_key(model_name=model_name, dataset_name=dataset_name)
+    task_type = MODEL_WRAPPER_REGISTRY.task_type_map[key]
+
+    # print("****",EVAL_WRAPPER_REGISTRY._registry_dict.keys())
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/wrappers/wrapper.py:93]: plz remove commented lines

File: deeplite_torch_zoo/utils/registry.py [hunk 12]
```diff
@@ -75,3 +75,31 @@ def get(self, model_name, dataset_name):
...
+            cls_name = self._registry_key(task_type=task_type,model_name=cls_name, dataset_name=dataset_name)
+            _register(cls_name, obj)
+            return obj
+
+        return wrap
+
+    def get(self, task_type, model_name=None, dataset_name=None):
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/utils/registry.py:100]: get method can take model_name and dataset_name, then you can determine model_type and dataset_type based on these (u can introduce dicts/methods for that in the eval registry) then use model_type + dataset_type to construct a registry key

**Review by ivan-lazarevich — CHANGES_REQUESTED** [review]

File: deeplite_torch_zoo/utils/registry.py (hunk 12 above)
Comment (ivan-lazarevich) [reply deeplite_torch_zoo/utils/registry.py:100]: it can be actually shared with the datasplit wrapper registry. instead of using `normalize_model_name`

File: deeplite_torch_zoo/utils/registry.py [hunk 13]
```diff
@@ -3,7 +3,47 @@
 
 from collections import namedtuple
 
-
+def normalize_model_name(net):
+    if "yolo" in net:
```
Comment (ivan-lazarevich) [inline deeplite_torch_zoo/utils/registry.py:7]: please refactor similar to this https://github.com/Deeplite/deeplite-torch-zoo/pull/151/files#diff-11d973c1c9bb7e814aa6854772bad44e42a08004601ca86cdffa9c137df01fe4R31

**Comment by ivan-lazarevich** [comment]
Looks good, thanks!
